include/nbl/builtin/hlsl/impl/tgmath_impl.hlsl

```diff
@@ -0,0 +1,767 @@
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<FloatingPoint>)
+struct isinf_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<FloatingPoint>) >
+{
+	static bool __call(NBL_CONST_REF_ARG(FloatingPoint) x)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::isInf<FloatingPoint>(x);
+#else
+		// GCC and Clang will always return false with call to std::isinf when fast math is enabled,
+		// this implementation will always return appropriate output regardless is fas math is enabled or not
+		using AsUint = typename unsigned_integer_of_size<sizeof(FloatingPoint)>::type;
+		return tgmath_impl::isinf_uint_impl(reinterpret_cast<const AsUint&>(x));
+#endif
+	}
+};
+
+template<typename V>
+NBL_PARTIAL_REQ_TOP(concepts::Vectorial<V>)
+struct isinf_helper<V NBL_PARTIAL_REQ_BOT(concepts::Vectorial<V>) >
+{
+	using output_t = vector<bool, hlsl::vector_traits<V>::Dimension>;
+
+	static output_t __call(NBL_CONST_REF_ARG(V) x)
+	{
+		using traits = hlsl::vector_traits<V>;
+		array_get<V, typename traits::scalar_type> getter;
+		array_set<output_t, typename traits::scalar_type> setter;
+
+		output_t output;
+		for (uint32_t i = 0; i < traits::Dimension; ++i)
+			setter(output, i, isinf_helper<typename traits::scalar_type>::__call(getter(x, i)));
+
+		return output;
+	}
+};
+
+// FLOOR
+
+template<typename V NBL_STRUCT_CONSTRAINABLE>
+struct floor_helper;
+
+template<typename FloatingPoint>
+NBL_PARTIAL_REQ_TOP(hlsl::is_floating_point_v<FloatingPoint> && hlsl::is_scalar_v<FloatingPoint>)
+struct floor_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(hlsl::is_floating_point_v<FloatingPoint> && hlsl::is_scalar_v<FloatingPoint>) >
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::floor(val);
+#else
+		return std::floor(val);
+#endif
+	}
+};
+
+template<typename V>
+NBL_PARTIAL_REQ_TOP(concepts::Vectorial<V>)
+struct floor_helper<V NBL_PARTIAL_REQ_BOT(concepts::Vectorial<V>) >
+{
+	static V __call(NBL_CONST_REF_ARG(V) vec)
+	{
+		using traits = hlsl::vector_traits<V>;
+		array_get<V, typename traits::scalar_type> getter;
+		array_set<V, typename traits::scalar_type> setter;
+
+		V output;
+		for (uint32_t i = 0; i < traits::Dimension; ++i)
+			setter(output, i, floor_helper<typename traits::scalar_type>::__call(getter(vec, i)));
+
+		return output;
+	}
+};
+
+// ERF
+
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct erf_helper;
+
+template<typename FloatingPoint>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<FloatingPoint>)
+struct erf_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<FloatingPoint>) >
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) _x)
+	{
+#ifdef __HLSL_VERSION
+		const FloatingPoint a1 = 0.254829592;
+		const FloatingPoint a2 = -0.284496736;
+		const FloatingPoint a3 = 1.421413741;
+		const FloatingPoint a4 = -1.453152027;
+		const FloatingPoint a5 = 1.061405429;
+		const FloatingPoint p = 0.3275911;
```

Comment:
pay attention to 2 things:
1. whether the constant literal has all the digits needed (float32 and float64 often have different digits
2. to get true FP64 constants in HLSL you need to use a stupid suffix, there's actually an `NBL_` macro you need to place around them cause DXC is holding back on implementing explicit sized fp16, fp32, fp64 suffices on float literals


Also, usually an implementation of a complex math function is different for float16, float32 and float64 (different levels of polynomial approximation/coefficients and number of steps/iterations) so I'd have a separate partial spec in preprocessor blocks of `__HLSL_VERSION